src/Middleware/RequestDecompression/src/RequestDecompressionBody.cs

```diff
@@ -0,0 +1,152 @@
+    {
+        _context = context;
+        _provider = provider;
+        _innerStream = context.Request.Body;
+    }
+
+    public override bool CanRead => true;
+
+    public override bool CanSeek => false;
+
+    public override bool CanWrite => _innerStream.CanWrite;
+
+    public override long Length
+    {
+        get { throw new NotSupportedException(); }
+    }
+
+    public override long Position
+    {
+        get { throw new NotSupportedException(); }
+        set { throw new NotSupportedException(); }
+    }
+
+    public Stream Stream => this;
+
+    public override int Read(byte[] buffer, int offset, int count)
+    {
+        OnRead();
+
+        if (_decompressionStream != null)
+        {
+            return _decompressionStream.Read(buffer, offset, count);
+        }
+
+        return _innerStream.Read(buffer, offset, count);
+    }
+
+    public override long Seek(long offset, SeekOrigin origin)
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void SetLength(long value)
+    {
+        throw new NotSupportedException();
+    }
+
+    public override void Flush()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void Write(byte[] buffer, int offset, int count)
+    {
+        throw new NotImplementedException();
+    }
+
+    public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
+        => TaskToApm.Begin(ReadAsync(buffer, offset, count, CancellationToken.None), callback, state);
+
+    public override int EndRead(IAsyncResult asyncResult)
+        => TaskToApm.End<int>(asyncResult);
+
+    public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        => await ReadAsync(buffer.AsMemory(offset, count), cancellationToken);
+
+    public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+    {
+        OnRead();
+
+        if (_decompressionStream != null)
+        {
+            return await _decompressionStream.ReadAsync(buffer, cancellationToken);
+        }
+
+        return await _innerStream.ReadAsync(buffer, cancellationToken);
+    }
+
+    private void OnRead()
+    {
+        if (!_decompressionChecked)
+        {
+            _decompressionChecked = true;
+
+            var decompressionProvider = ResolveDecompressionProvider();
+
+            if (decompressionProvider != null)
+            {
+                _decompressionStream = decompressionProvider.CreateStream(_innerStream);
+            }
+        }
+    }
+
+    private IDecompressionProvider? ResolveDecompressionProvider()
+    {
+        if (!_providerCreated)
+        {
+            _providerCreated = true;
+            _decompressionProvider = _provider.GetDecompressionProvider(_context);
+        }
+
+        return _decompressionProvider;
+    }
```

Comment:
Request decompression probably doesn't need this lazy resolution, this logic can happen eagerly in the middleware.

Response compression needed to use this approach because it didn't know what the response would be when the middleware ran.

Reply:
Ah, good to know. I've updated the `RequestDecompressionBody` to eagerly resolve the decompression provider and decompression stream in its constructor.